File: pm-backend/src/main/kotlin/io/vliet/plusmin/domain/Rekening.kt
package io.vliet.plusmin.domain

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonInclude
import jakarta.persistence.*
import java.math.BigDecimal

@Entity
@Table(
    name = "rekening",
    uniqueConstraints = [UniqueConstraint(columnNames = ["gebruiker", "naam"])]
)
class Rekening(
    @Id
    @GeneratedValue(generator = "hibernate_sequence", strategy = GenerationType.SEQUENCE)
    @SequenceGenerator(
        name = "hibernate_sequence",
        sequenceName = "hibernate_sequence",
        allocationSize = 1
    )
    val id: Long = 0,
    val naam: String,
    @ManyToOne
    @JsonIgnore
    @JoinColumn(name = "gebruiker_id")
    val gebruiker: Gebruiker,
    @Enumerated(EnumType.STRING)
    val rekeningSoort: RekeningSoort,
    val nummer: String? = null,
    @JsonInclude(JsonInclude.Include.NON_NULL)
    val bankNaam: String? = null,
    val sortOrder: Int,
    @OneToMany(mappedBy = "rekening", fetch = FetchType.EAGER, cascade = [CascadeType.ALL], orphanRemoval = true)
    var budgetten: List<Budget> = emptyList()
) {
    companion object {
        val sortableFields = setOf("id", "naam", "afkorting")
    }

    fun fullCopy(
        naam: String = this.naam,
        gebruiker: Gebruiker = this.gebruiker,
        rekeningSoort: RekeningSoort = this.rekeningSoort,
        nummer: String? = this.nummer,
        bankNaam: String? = this.bankNaam,
        sortOrder: Int = this.sortOrder,
        budgetten: List<Budget> = this.budgetten
    ) = Rekening(this.id, naam, gebruiker, rekeningSoort, nummer, bankNaam, sortOrder, budgetten)

    data class RekeningDTO(
        val id: Long = 0,
        val naam: String,
        val rekeningSoort: String,
        val nummer: String?,
        val bankNaam: String?,
        val saldo: BigDecimal = BigDecimal(0),
        val sortOrder: Int,
        val budgetten: List<Budget>? = emptyList()
    )

    fun toDTO(): RekeningDTO {
        return RekeningDTO(
            this.id,
            this.naam,
            this.rekeningSoort.toString(),
            this.nummer,
            this.bankNaam,
            sortOrder = this.sortOrder,
            budgetten = this.budgetten
        )
    }

    enum class RekeningSoort {
        BETAALREKENING, SPAARREKENING, CONTANT, CREDITCARD, AFLOSSING, RESERVERING,
        INKOMSTEN, RENTE, UITGAVEN
    }
}

val resultaatRekeningSoort = arrayOf(
    Rekening.RekeningSoort.INKOMSTEN,
    Rekening.RekeningSoort.RENTE,
    Rekening.RekeningSoort.UITGAVEN,
)

val balansRekeningSoort = arrayOf(
    Rekening.RekeningSoort.BETAALREKENING,
    Rekening.RekeningSoort.SPAARREKENING,
    Rekening.RekeningSoort.CONTANT,
    Rekening.RekeningSoort.CREDITCARD,
    Rekening.RekeningSoort.AFLOSSING,
    Rekening.RekeningSoort.RESERVERING
)

File: pm-frontend/src/model/Rekening.ts
import { BetalingDTO } from "./Betaling";
import { Budget } from "./Budget";

export type Rekening = {
    id: number;
    naam: string;
    rekeningSoort: RekeningSoort;
    nummer: string | undefined;
    sortOrder: number;
    budgetten: Budget[];
}

export enum RekeningSoort {
    inkomsten = 'INKOMSTEN',
    uitgaven = 'UITGAVEN',
    betaalrekening = 'BETAALREKENING',
    spaarrekening = 'SPAARREKENING',
    contant = 'CONTANT',
    creditcard = 'CREDITCARD',
    aflossing = 'AFLOSSING',
    reservering = 'RESERVERING'
}

export type RekeningSoortPaar = {
    bron: RekeningSoort[];
    bestemming: RekeningSoort[];
}

export type RekeningPaar = {
    bron: Rekening[];
    bestemming: Rekening[];
}

export const balansRekeningSoorten: RekeningSoort[] = [
    RekeningSoort.betaalrekening,
    RekeningSoort.spaarrekening,
    RekeningSoort.contant,
    RekeningSoort.creditcard,
    RekeningSoort.aflossing,
    RekeningSoort.reservering];

export const resultaatRekeningSoorten = [
    RekeningSoort.inkomsten,
    RekeningSoort.uitgaven];

export const aflossenRekeningSoorten = [
    RekeningSoort.creditcard,
    RekeningSoort.aflossing];

export const reserverenRekeningSoorten = [
    RekeningSoort.reservering];

export const betaalmethodeRekeningSoorten = [
    RekeningSoort.betaalrekening,
    RekeningSoort.contant,
    RekeningSoort.creditcard,
]

export const inkomstenRekeningSoorten = [
    RekeningSoort.inkomsten,
]

export const uitgavenRekeningSoorten = [
    RekeningSoort.uitgaven,
    RekeningSoort.aflossing,
    RekeningSoort.reservering,
]

export const interneRekeningSoorten = [ 
    RekeningSoort.spaarrekening,
    RekeningSoort.contant,
    RekeningSoort.creditcard,
    RekeningSoort.reservering,
]

export const cashflowRekeningSoorten = [
    RekeningSoort.betaalrekening,
    RekeningSoort.spaarrekening,
    RekeningSoort.contant,
    RekeningSoort.creditcard,
]

export const berekenBedragVoorRekenining = (betaling: BetalingDTO, rekening: Rekening | undefined) => {
    if (rekening === undefined) return betaling.bedrag // filter = 'all'
    const factor = resultaatRekeningSoorten.includes(rekening.rekeningSoort) ? -1 : 1
    if (betaling.bron === rekening.naam) return -betaling.bedrag * factor
    if (betaling.bestemming === rekening.naam) return betaling.bedrag * factor
    return 0
  }
  